### PR TITLE
Simplify sdks/java/harness shadowJar

### PR DESCRIPTION
**Pull request by apilloud**
This reduces the surface area of the shadowJar to avoid conflicts with user code. It is expected that the user will provide a working copy of Beam core and its dependencies.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by apilloud** [comment]
Failure is #24267

**Comment by apilloud** [comment]
Run SQL PreCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -1022,6 +1022,50 @@ class BeamModulePlugin implements Plugin<Project> {
         }
       }
 
+      // Ban these dependencies from all configurations
+      project.configurations.all {
```
Comment (lukecwik) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:1026]: Why did you need/want to move these up to here?

File: sdks/java/harness/build.gradle [hunk 2]
```diff
@@ -18,35 +18,44 @@
...
   testShadowJar: true,
+  shadowJarValidationExcludes: [
+    "io/github/classgraph/**",
+    "nonapi/io/github/classgraph/**",
+    "org/apache/beam/fn/**",
+    "org/apache/beam/model/**",
+    "org/apache/beam/runners/**",
+    "org/apache/beam/sdk/**",
+    "org/checkerframework/**",
+    "org/github/jamm/**",
```
Comment (lukecwik) [inline sdks/java/harness/build.gradle:45]: I think jamm should be provided as it needs to be on the boot agent classpath

File: sdks/java/harness/build.gradle [hunk 3]
```diff
@@ -55,21 +64,19 @@ description = "Apache Beam :: SDKs :: Java :: Harness"
...
 dependencies {
-  dependOnShadedProjects.each {
-    implementation project(path: it, configuration: "shadow")
-  }
-  dependOnProjects.each {
-    implementation project(it)
-  }
-  shadow library.java.vendored_guava_26_0_jre
+  implementation project(path: ":model:fn-execution", configuration: "shadow")
+  implementation project(path: ":model:pipeline", configuration: "shadow")
```
Comment (lukecwik) [inline sdks/java/harness/build.gradle:68]: I would expect this to be provided

File: sdks/java/harness/build.gradle [hunk 4]
```diff
@@ -55,21 +64,19 @@ description = "Apache Beam :: SDKs :: Java :: Harness"
...
+  implementation project(":runners:core-construction-java")
+  implementation project(":runners:core-java")
+  implementation project(":sdks:java:fn-execution")
```
Comment (lukecwik) [inline sdks/java/harness/build.gradle:71]: I would expect this to be provided

File: sdks/java/harness/build.gradle [hunk 5]
```diff
@@ -18,35 +18,44 @@
...
   automaticModuleName: 'org.apache.beam.fn.harness',
-  validateShadowJar: false,
   testShadowJar: true,
+  shadowJarValidationExcludes: [
+    "io/github/classgraph/**",
+    "nonapi/io/github/classgraph/**",
+    "org/apache/beam/fn/**",
+    "org/apache/beam/model/**",
+    "org/apache/beam/runners/**",
+    "org/apache/beam/sdk/**",
```
Comment (lukecwik) [inline sdks/java/harness/build.gradle:43]: can this be narrowed down to `org/apache/beam/sdk/fn/**`?

File: sdks/java/harness/build.gradle [hunk 6]
```diff
@@ -18,35 +18,44 @@
...
   ],
   automaticModuleName: 'org.apache.beam.fn.harness',
-  validateShadowJar: false,
   testShadowJar: true,
+  shadowJarValidationExcludes: [
+    "io/github/classgraph/**",
+    "nonapi/io/github/classgraph/**",
+    "org/apache/beam/fn/**",
+    "org/apache/beam/model/**",
+    "org/apache/beam/runners/**",
```
Comment (lukecwik) [inline sdks/java/harness/build.gradle:42]: Can this be narrowed down to the two underlying packages

File: sdks/java/harness/build.gradle [hunk 7]
```diff
@@ -18,35 +18,44 @@
...
     'AssignWindowsRunner': 'https://github.com/typetools/checker-framework/issues/3794',
     'WindowMergingFnRunner': 'https://github.com/typetools/checker-framework/issues/3794',
   ],
   automaticModuleName: 'org.apache.beam.fn.harness',
-  validateShadowJar: false,
   testShadowJar: true,
+  shadowJarValidationExcludes: [
+    "io/github/classgraph/**",
+    "nonapi/io/github/classgraph/**",
+    "org/apache/beam/fn/**",
```
Comment (lukecwik) [inline sdks/java/harness/build.gradle:40]: Where does this one come from?

**Comment by apilloud** [comment]
This still fails a bunch of tests, not yet ready for review.

File: sdks/java/harness/build.gradle (hunk 5 above)
Comment (apilloud) [reply sdks/java/harness/build.gradle:43]: Yes, done.

File: sdks/java/harness/build.gradle (hunk 6 above)
Comment (apilloud) [reply sdks/java/harness/build.gradle:42]: Narrowed down to `"org/apache/beam/fn/core/**"`

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 1 above)
Comment (apilloud) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:1026]: `Cannot change dependencies of dependency configuration ':sdks:java:harness:provided' after it has been resolved.` This changes all configurations.

**Comment by apilloud** [comment]
Spotless failure is #20819

**Comment by apilloud** [comment]
Run Spotless PreCommit

File: sdks/java/harness/build.gradle (hunk 3 above)
Comment (apilloud) [reply sdks/java/harness/build.gradle:67]: I've moved all transitive dependencies of `:sdks:java:core` to provided.

File: sdks/java/harness/build.gradle (hunk 4 above)
Comment (apilloud) [reply sdks/java/harness/build.gradle]: Does this need a CHANGES.md update? Who should be providing it?

File: sdks/java/harness/build.gradle (hunk 2 above)
Comment (apilloud) [reply sdks/java/harness/build.gradle]: Done.

**Comment by apilloud** [comment]
Run Java PreCommit

**Comment by apilloud** [comment]
Run Java PreCommit

**Comment by apilloud** [comment]
R: @lukecwik

File: sdks/java/harness/build.gradle (hunk 4 above)
Comment (lukecwik) [reply sdks/java/harness/build.gradle]: Was wrong, we'll need to keep it as is.

**Review by lukecwik — CHANGES_REQUESTED** [review]

File: sdks/java/harness/build.gradle (hunk 4 above)
Comment (apilloud) [reply sdks/java/harness/build.gradle]: Reverted.

File: sdks/java/harness/build.gradle (hunk 7 above)
Comment (apilloud) [reply sdks/java/harness/build.gradle:50]: Done

File: CHANGES.md [hunk 8]
```diff
@@ -78,6 +78,8 @@
   dependency of the Java SDK Harness. Some users of a portable runner (such as Dataflow Runner v2)
   may have an undeclared dependency on this package (for example using GCS with
   TextIO) and will now need to declare the dependency.
+* `beam-sdks-java-core` is no longer a dependency of the Java SDK Harness. Users of a portable
+  runner (such as Dataflow Runner v2) will need to provide this package and it's dependecies.
```
[the remark this answers is not in the document]
Comment (apilloud) [reply CHANGES.md:82]: Done.

File: sdks/java/harness/build.gradle [hunk 9]
```diff
@@ -18,35 +18,51 @@
...
-// and also the set of project level dependencies.
-def dependOnShadedProjects = [":model:pipeline", ":model:fn-execution", ":sdks:java:core"]
-def dependOnProjects = [":sdks:java:fn-execution",
-                        ":runners:core-java", ":runners:core-construction-java"]
+configurations {
+  provided
+}
+
+dependencies {
+  // provided dependencies are declared early so they can be resolved in shadowClosure
```
[the remark this answers is not in the document]
Comment (apilloud) [reply sdks/java/harness/build.gradle:26]: Done.

File: sdks/java/harness/build.gradle [hunk 10]
```diff
@@ -18,35 +18,51 @@
...
 plugins { id 'org.apache.beam.module' }
 
-// We specifically enumerate all the projects that we depend on since
-// the list is used in both defining the included set for the uber jar
-// and also the set of project level dependencies.
-def dependOnShadedProjects = [":model:pipeline", ":model:fn-execution", ":sdks:java:core"]
-def dependOnProjects = [":sdks:java:fn-execution",
-                        ":runners:core-java", ":runners:core-construction-java"]
+configurations {
+  provided
```
[the remark this answers is not in the document]
Comment (apilloud) [reply sdks/java/harness/build.gradle:22]: I'm assuming your goal is to revert the changes to `BeamModulePlugin.groovy`? That is not possible. I did prototype this change and got the same error as `project.configurations.all` applies to all configurations not just ones defined in the file.

File: sdks/java/harness/build.gradle [hunk 11]
```diff
@@ -18,35 +18,51 @@
...
+  provided project(path: ":model:pipeline", configuration: "shadow")
+  provided project(path: ":sdks:java:core", configuration: "shadow")
+  provided library.java.joda_time
+  provided library.java.slf4j_api
+  provided library.java.vendored_grpc_1_48_1
+  provided library.java.vendored_guava_26_0_jre
+
+  // :sdks:java:fn-execution and transitive dependencies
+  provided project(path: ":model:fn-execution", configuration: "shadow")
+  provided project(":sdks:java:fn-execution")
```
[the remark this answers is not in the document]
Comment (apilloud) [reply sdks/java/harness/build.gradle]: Done.

File: sdks/java/harness/build.gradle (hunk 10 above)
[the remark this answers is not in the document]
Comment (lukecwik) [reply sdks/java/harness/build.gradle:22]: That was the primary goal. We can leave as is.

File: sdks/java/harness/jmh/build.gradle [hunk 12]
```diff
@@ -41,6 +41,7 @@ dependencies {
     implementation library.java.joda_time
     runtimeOnly library.java.slf4j_jdk14
     jammAgent library.java.jamm
+    runtimeOnly library.java.jamm
```
Comment (lukecwik) [inline sdks/java/harness/jmh/build.gradle:44]: testRuntimeOnly?

**Review by lukecwik — APPROVED** [review]

**Comment by apilloud** [comment]
Failure in precommit is #24383

**Comment by apilloud** [comment]
Run Java PreCommit

**Comment by apilloud** [comment]
Failure in precommits is #24384 and #21714

**Comment by apilloud** [comment]
Run Java PreCommit

**Comment by apilloud** [comment]
Failures are #24267 and #24389

**Comment by apilloud** [comment]
Run Java PreCommit
